File: go/algorithms/shell.go
package algorithms

// ShellSort sorts in-place using Shell's method with gap sequence halving.
func ShellSort(arr []int) {
    for gap := len(arr) / 2; gap > 0; gap /= 2 {
        for i := gap; i < len(arr); i++ {
            temp := arr[i]
            j := i
            for j >= gap && arr[j-gap] > temp {
                arr[j] = arr[j-gap]
                j -= gap
            }
            arr[j] = temp
        }
    }
}

File: rust/src/shell.rs
pub fn shell_sort(arr: &mut [i32]) {
    let mut gap = arr.len() / 2;
    while gap > 0 {
        for i in gap..arr.len() {
            let temp = arr[i];
            let mut j = i;
            while j >= gap && arr[j - gap] > temp {
                arr[j] = arr[j - gap];
                j -= gap;
            }
            arr[j] = temp;
        }
        gap /= 2;
    }
}
